### PR TITLE
Add a registry rebuild if drush updb pukes

### PR DESCRIPTION
**Pull request by jbickar**
# READY FOR REVIEW

# Summary
- TL;DR - what's this PR for? Rebuild the registry if our first `drush updb` fails

# Needed By (Date) Before we start migrating at scale

# Criticality
- How critical is this PR on a 1-10 scale? 6

# Steps to Test

1. Check out this branch
2. Add `jumpstart-sts vhost="sts"` to an inventory file (the group doesn't matter, you can add it in Earth or GSE from the last testing you did).
3. Clone to `dev`
4. Get a fatal class not defined error
5. Update your inventory entry to be `jumpstart-sts vhost="sts" ignore_updb_errors="TRUE"`
6. Re-run the migration
7. See it fail on the first `drush updb`, ignore the error, and then the second succeeds because we rebuilt the registry.

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules? ACSF/Sites 2.0

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: roles/upload-site/tasks/main.yml [hunk 1]
```diff
@@ -41,20 +41,21 @@
 # per-site basis. That will allow us to move past errors in this task, and run
 # "drush updb" a second time in the following task.
 - name: Update database schema with drush updb
-  shell: "{{drush_alias }} {{ item }}"
+  shell: "{{ drush_alias }} {{ item }}"
   with_items:
     - "-y updb"
+    - "-y rr"
```
Comment (jbickar) [inline roles/upload-site/tasks/main.yml]: This is the only functional change. The other ones are just syntax cleanup.

Comment (sherakama) [inline roles/upload-site/tasks/main.yml]: Won't this run rr on every site not just the failed ones?

File: roles/upload-site/tasks/main.yml [hunk 2]
```diff
@@ -41,20 +41,21 @@
...
   ignore_errors: "{% if ignore_updb_errors == 'TRUE' %}yes{% endif %}"
 
 # This only runs "drush updb" if ignore_updb_errors == "TRUE"
 - name: Run drush updb a second time for recalcitrant sites
-  shell: "{{drush_alias }} {{ item }}"
+  shell: "{{ drush_alias }} {{ item }}"
   with_items:
     - "-y updb"
```
Comment (sherakama) [inline roles/upload-site/tasks/main.yml]: Don't you want to add it before this one?

File: roles/upload-site/tasks/main.yml (hunk 1 above)
Comment (jbickar) [reply roles/upload-site/tasks/main.yml]: Right. Let’s move that to the next task.

Comment (jbickar) [reply roles/upload-site/tasks/main.yml]: Before the second updb

**Comment by jbickar** [comment]
Let me do some testing. There may be some order-of-operations issues.

**Comment by sherakama** [comment]
Whoops, I just saw your comment. I will hold off.

**Comment by sherakama** [comment]
Also, it might be better to truncate the registry tables (registry, registry_file) prior to importing the DB, so that you don't have to bootstrap Drupal to run `drush rr` once it is on the new environment.

**Comment by jbickar** [comment]
I tried that in troubleshooting the `supri-b` "missing column" issue, and Drupal doesn't like those tables to be empty.

**Comment by sherakama** [comment]
🤔 registry_rebuild only bootstraps to `DRUPAL_BOOTSTRAP_CONFIGURATION` so that actually may be useful for every site prior to running drush updb

**Comment by jbickar** [comment]
We do run registry-rebuild for all sites later in the process.

**Comment by sherakama** [comment]
But you want it before, to tell drupal where all the modules have moved to prior to running updb, otherwise it may miss or not load something.

**Comment by jbickar** [comment]
That's what I'm thinkin'.

**Comment by jbickar** [comment]
Here's what's happening with `ds_jumpstart-sts`:

1. The first `drush updb` fails with "Error: Class 'FieldCollectionItemEntity' not found in\n/mnt/www/html/cardinald702test/docroot/sites/all/modules/stanford/stanford_landing_page/stanford_landing_page.module,". OK, fine, we're ignoring errors on that task.
2. `drush rr` fails with "Unknown column     [error]", "'base.status' in 'field list", but it actually _succeeds_ in rebuilding the registry.
3. The subsequent `drush updb` succeeds.

So I think we need the `rr` in a task where errors get ignored. Maybe I'll add it to its own task that only gets run if `ignore_updb_errors="TRUE"`.

There's a lot of crazy OOO going on here.

**Comment by jbickar** [comment]
OK, give the latest a whirl with:
```
jumpstart-sts vhost="sts" ignore_updb_errors="TRUE"
supri-b vhost="supri-b" ignore_updb_errors="TRUE"
```

File: roles/upload-site/tasks/main.yml [hunk 3]
```diff
@@ -41,32 +41,40 @@
...
   when: ignore_updb_errors == "TRUE"
 
 - name: Set site up as a Department site
-  shell: "{{drush_alias }} {{ item }}"
+  shell: "{{ drush_alias }} {{ item }}"
   with_items:
-    - "-y vset install_profile stanford_dept"
+    - "-y vset install_profile stanford_sites_jumpstart"
```
Comment (sherakama) [inline roles/upload-site/tasks/main.yml]: How come jumpstart? Also, it is time to make this dynamic?

Comment (jbickar) [reply roles/upload-site/tasks/main.yml]: Damn these ham hands! I didn’t mean to commit this.  And yes it is, but I got distracted by something shiny.

**Comment by sherakama** [comment]
@jbickar,

Ok, I have added a new product group and task where it looks for the helper modules and sets the installation profile accordingly. Please have a review.

File: roles/upload-site/tasks/main.yml [hunk 4]
```diff
@@ -41,32 +41,52 @@
...
 # This only runs "drush updb" if ignore_updb_errors == "TRUE"
 - name: Run drush updb a second time for recalcitrant sites
-  shell: "{{drush_alias }} {{ item }}"
+  shell: "{{ drush_alias }} {{ item }}"
   with_items:
     - "-y updb"
   when: ignore_updb_errors == "TRUE"
 
+# Set the install profile by looking for the enabled supporting module.
+- name: Find and set the install profile by helper modules.
```
Comment (sherakama) [inline roles/upload-site/tasks/main.yml]: New function for products. Used swsbuildjsa as an example site to migrate.

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: https://swsbuildjsa-dev.cardinalsites.stanford.edu/

Comment (jbickar) [reply roles/upload-site/tasks/main.yml]: Can we assume a 1:1 here? Or shall we just assume a 1:1 gets us "close enough"?

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: 1:1 is pretty accurate. If the module is available and enabled (which is what module_exists checks) then we can be almost certain that this site is what it is we think it is.

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: 🤔 hmm.. I didn't think that we had anything that would be JSA without the helper module. That would throw a wrench into this mix.

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: Initial commit May 1, 2014 for JSA

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: Initial commit Dec 13, 2013 for JSA profile.

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: Well.... shoot.

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: Yeah, jsa-content would stay at the stanford profile... which is right imho as it really isn't a JSA.

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: English would be set to stanford_jumpstart however...

Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: Ok, anything deployed after May 1st, 2014 can get the products group. Anything before... well... The likely need special treatment anyhow.

Comment (jbickar) [reply roles/upload-site/tasks/main.yml]: Yeah that sounds about right. English was our build site for JSA, so by definition it wouldn't have `stanford_jumpstart_academic`. Art would be the other one; it only has `stanford_jumpstart` enabled but I *think* it's a JSA? 🤔

Comment (jbickar) [reply roles/upload-site/tasks/main.yml]: 1:1 gets us "close enough". We know that we have some ❄️ ❄️ ❄️

Comment (jbickar) [reply roles/upload-site/tasks/main.yml]: The other approach to this that I was thinking of was to set it as a variable on a per-site basis (or put the sites into groups with an `install_profile` variable), rather than auto-detect. But since you've already done the work for auto-detect, and my approach would be more work. And I don't like work.

**Comment by jbickar** [comment]
Scope crept a little bit on this one 😂 

I tested with `supri-b` set to a dept site and `jumpstart-sts` set to JSA.

```
PLAY RECAP *********************************************************************************************************************
jumpstart-sts              : ok=61   changed=29   unreachable=0    failed=0   
supri-b                    : ok=61   changed=28   unreachable=0    failed=0   

jbickar@vagrant-ubuntu-trusty-64:~/shared/ansible-playbooks$ drush @acsf.test.cardinald7.supri-b vget install_pr
install_profile: stanford_dept

jbickar@vagrant-ubuntu-trusty-64:~/shared/ansible-playbooks$ drush @acsf.test.cardinald7.jumpstart-sts vget install_pr
install_profile: stanford_sites_jumpstart_academic

jbickar@vagrant-ubuntu-trusty-64:~/shared/ansible-playbooks$ drush @acsf.test.cardinald7.jumpstart-sts pmi stanford_jumpstart_academic
 Extension        :  stanford_jumpstart_academic                               
 Project          :  stanford_jumpstart_academic                               
 Type             :  module                                                    
 Title            :  Stanford Jumpstart Academic                               
 Description      :  For All things custom with the Jumpstart Academic Product 
 Version          :  7.x-5.1-alpha5                                            
 Package          :  Other                                                     
 Core             :  7.x                                                       
 PHP              :  5.2.4                                                     
 Status           :  enabled                                                   
 Path             :  sites/all/modules/stanford/stanford_jumpstart_academic    
 Schema version   :  7100                                                      
 Files            :  none                                                      
 Requires         :  stanford_jumpstart                                        
 Required by      :  stanford_sites_jumpstart_academic                         
 Permissions      :  none                                                      
 Configure        :  None                                                      

jbickar@vagrant-ubuntu-trusty-64:~/shared/ansible-playbooks$ drush @acsf.test.cardinald7.supri-b pmi stanford_jumpstart_academic 
stanford_jumpstart_academic was not found.                                                                             [warning]
```

Both look good too:
- https://sts-test.cardinalsites.stanford.edu
- https://supri-b-test.cardinalsites.stanford.edu

File: roles/upload-site/tasks/main.yml (hunk 4 above)
Comment (sherakama) [reply roles/upload-site/tasks/main.yml]: Yeah, I think we can throw some spaghetti on the wall with this function and see what sticks. For the stuff that drops be a bit more explicit and run it with a set variable.
